File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int minr = 0;
        int minc = 0;
        int maxr = matrix.length-1;
        int maxc = matrix[0].length-1;

        int count = 0;
        int total = matrix.length * matrix[0].length;
        
        List<Integer> arr = new ArrayList<>();

        while(count < total){

            //topwall
            for(int i = minr, j = minc; j <= maxc && count < total; j++){
                  arr.add(matrix[i][j]);
                  count++;
            }
            minr++;

            // right wall
            for(int i = minr, j = maxc; i <= maxr && count < total; i++){
                arr.add(matrix[i][j]);
                count++;
            }
            maxc--;

            // bottom wall
            for(int i = maxr, j = maxc; j >= minc && count < total; j--){
                arr.add(matrix[i][j]);
                count++;
            }
            maxr--;

            // left wall
            for(int i = maxr, j = minc; i >= minr && count < total; i--){
                arr.add(matrix[i][j]);
                count++;
            }
            minc++;
        }

        return arr;

    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        spiral = []

        minr = 0
        minc = 0
        maxr = len(matrix)-1
        maxc = len(matrix[0])-1
        
        count = 0
        total = len(matrix) * len(matrix[0])
        
        while count < total:
            
            #Top wall
            for i in range(minc,maxc+1):
                spiral.append(matrix[minr][i])
                count+=1
            
            minr+=1


            # right wall
            for i in range(minr,maxr+1):
                spiral.append(matrix[i][maxc])
                count+=1
            
            maxc -= 1

            #bottom wall
            if minr <= maxr:

                for i in range(maxc,minc-1,-1):
                    spiral.append(matrix[maxr][i])
                    count+=1
                maxr -= 1

            #left wall
            if minc <= maxc:
                for i in range(maxr,minr-1,-1):
                    spiral.append(matrix[i][minc])
                    count+=1
            
                minc += 1
        
        return spiral